File: Tail Call Optimization/proposal/code/list.hs
data List a = Empty | Cons a (List a)

howMany :: List a -> Int
howMany list = size 0 list
	where size n []     = n
	      size n (x:xs) = size (n + 1) xs

main :: Int -> Bool
main n = let list = ... -- create list of size $n$
         in howMany list == n

File: Tail Call Optimization/report/code/list.java
abstract class List<T> {
	int howMany() { return size(0); }
	abstract int size(int i);
}

class Empty extends List<T> {
	int size(int i) { return i; }
}

class Cons extends List<T> {
	T element;
	List<T> rest;
	int size(int i) { return rest.size(i + 1); }
}